### DISCUSSION TITLE
Lesson 12 Stable Coin Test Invisible In Coverage

### DISCUSSION THREAD

**Discussion by Niferu** [Cyfrin/foundry-full-course-cu#762] [Q&A] [answered]
Hello guys,

I have encountered some strange issue. I have written some tests for my stable coin and all tests pass and are visible, but once I run `forge coverage` it shows 0% for my stable coin. Seems like some naming issue? I have no idea to be honest what is causing that behaviour.

Please see below screen with example:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/110537615/1e272f67-d9dd-4320-8614-ceb1652ca640)

Please also see my repository here: [link](https://github.com/Neftyr/DeFi-Stablecoin)

**Comment by shikhar229169** [comment]
Foundry will consider only those functions in coverage reports which are defined in the contract not the functions which are inherited, so if you have not tested the burn and mint function, it will show reports as 0%
So you need to make sure that you have tested the burn and mint function in your test otherwise it will show 0 in coverage.

**Reply by Niferu** [reply]
Hi @shikhar229169 

Hmm ok, but I have even copied DecentralizedStableCoin code from Patrick repo and copied tests for it also and foundry worked showing 100% coverage, so I believe that is not the case unfortunately.

**Reply by theirrationalone** [reply]
Patrick would have made an integration test which has mint & burn dsc from dscstablecoin, involved... That's why coverage was appearing 100%, i'm also surprised if branches were also covered???

**Reply by shikhar229169** [reply]
Yes, patrick tested everything:
Here you go: https://github.com/Cyfrin/foundry-defi-stablecoin-f23/blob/main/test/unit/DecentralizedStablecoinTest.t.sol

**Reply by Niferu** [reply]
As I have already mentioned I have only copied DecentralizedStableCoin test and DecentralizedStableCoin from repo and tests worked in my repo for those. My coin contract and test were the same and those are not visible for coverage for some reason I have not implemented any integration tests yet. This has to be something wrong with names or idk.. but definitely that is not what you mentioned above. I will try to solve it somehow and will keep you posted because if it is working for DSC it should work for my coin same way also.

**Reply by shikhar229169** [reply]
There is something wrong with foundry.
I cloned your repo and ran the tests and it was showing 0%, but then I added a function as you can see below in the images in your NeftyrStableCoin contract, and the coverage showed up for the contract.

Refer to the images:
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/111875681/28ff29f5-b304-4c28-9f74-ea7db0e4bd70)
![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/111875681/70a537ed-a686-4435-99e7-271aa3e1bffe)

@PatrickAlphaC help us.

**Reply by Niferu** [reply]
@shikhar229169 

Thanks a lot thats something at least! I knew that is something wrong with foundry or naming, but I'm still wondering why even with same code Patrick's coin test coverage is showing up normally. I will keep digging on that too

**Reply by shikhar229169** [reply]
@Neftyr You can make this issue on [foundry repo](https://github.com/foundry-rs/foundry/issues) if you want. 
Idk why it's happening, I went to my Stable Coin which I coded along with Patrick and the coverage is showing up.

**Reply by Niferu** [reply, marked as answer]
I got the issue!

"Where 2 contracts inherit the same interface and implement the same functions and have the same source layout, although they are both tested fully, forge coverage fails to recognise hits on one of the contracts.
In this case it would fail to recognise hits on SecondContract as FirstContract comes first alphabetically.

Confirmed this by changing their names to ensure it picked up the alphabetically first contract and also confirmed that when the source layout of either contracts is changed, forge coverage manages to recognise all hits on both of them."

This issue is reported here: [link](https://github.com/foundry-rs/foundry/issues/5729)

And it is obviously `foundry bug`

So in course repo it is working because D in before M and in my case it is not visible as N is after M in alphabet. Only file name matters not contract name.

Thanks @shikhar229169!

**Reply by shikhar229169** [reply]
What is 'M' referred to here?

**Reply by Niferu** [reply]
@shikhar229169 

I meant: In this project case N (NeftyrStableCoin) is after M (MockFailedMintNFR) aplhabetically, which means only Mock will be shown in coverage.

**Comment by theirrationalone** [comment]
Heyy @shikhar229169 , @Neftyr !, why foundry considers invariants tests scripts as contracts expected to be tested out.... Something like LinkToken mock. ??

**Reply by shikhar229169** [reply]
It takes every solidity file in the coverage.
But there is no point saying to write test to test the test, lol.
